### ISSUE TITLE
Split definitions cause problems for python importer

### ISSUE THREAD

**Issue by edreamleo** [leo-editor/leo-editor#354]
A new unit test attempts to import all of Leo's core files.  It fails for 7 files. The old importers likely would have failed as well. The following example, from leoGlobals.py, illustrates the problems:

```python
if isPython3: # g.not defined yet.

    def u(s):
        '''Return s, converted to unicode from Qt widgets.'''
        return s

    def ue(s, encoding):
        return s if g.isUnicode(s) else str(s, encoding)

else:

    def u(s):
        '''Return s, converted to unicode from Qt widgets.'''
        # Use builtins to suppress pyflakes complaint.
        # pylint: disable=no-member, undefined-variable
        return builtins.unicode(s) # Suppress pyflakes complaint.

    def ue(s, encoding):
        # Use builtins to suppress pyflakes complaint.
        # pylint: disable=no-member, undefined-variable
        return builtins.unicode(s, encoding) 
```

The importer can't use a single `@others` because the `else:` statement intervenes. Moreover, it can't use section references whose names are the def names because there are two definitions for both `e` and `ue`.

This looks like the acid test of the python importer. There may be other complications.  All must be fixed before we can call the importers complete.

**Comment by edreamleo** [comment]
The old importers fail only for leoTangle.py.  Some new invention is required.

**Comment by edreamleo** [comment]
**Code generation for old importers**

I have *temporarily  restored the old python importers in order to see how they are able to handle this issue successfully.  The result are interesting.

The old importers generate nodes only for what could be called "top-level" classes and defs *and* any nested classes and defs.  As a result, the old python importers generates the following in leoColorizer.py:

```
if QtGui:
    if python_qsh:
        base_highlighter = PythonQSyntaxHighlighter
    else:
        base_highlighter = QtGui.QSyntaxHighlighter

    class LeoQtSyntaxHighlighter(base_highlighter):
        ...
class QScintillaColorizer(ColorizerMixin):
    '''A colorizer for a QsciScintilla widget.'''
    @others
```
**Important**: The code above appears as "preamble" code in a node called QScintillaColorizer(!)  The LeoQtSyntaxHighlighter class is fairly long, and the old importer code would place it in the preamble no matter how long it might be.  This isn't a code generation model I want to emulate.

It's not completely clear what the old importers would do with split `@others` inside class/def nodes.  New unit tests would reveal the answer, but the answer might not be worth knowing.

**Summary**

The old importers work by "accident" in some sense.  They generate accurate code, but not optimal code.

The new importers will probably generate section references for overly-indented classes and defs.  It will be tricky, but new unit tests (based on all presently failing test files) will guide the development of new code.

**Comment by edreamleo** [comment]
Options for code generation are limited if section references are not allowed.  We could use the approach taken by the old code generators, or we could special case `if` and `else` statements. The new code generators have a new tool at their disposal: look-ahead patterns.

Here is the acid test, in my present thinking:
```
if aTest:
    class aClass:
        ... # Arbitrarily long class.
else:
    class aClass:
        ... # Arbitrarily long class.
```

Without section references, the only solution is to split this into two trees, like this:

```
+ if aTest:
    @others
       + class aTest
           @others
+ else:
    @others
       + class aTest
           @others
```

Lookahead patterns are important here.  We don't want to generate all these nodes unless the `if` or `else` clauses actually contain a class/def.

This is not likely to be easy, but lookahead patterns will simplify matters as much as can be expected.  I'll be working on this approach next.

**Comment by tbnorth** [comment]
I wouldn't think badly of Leo if only top level definitions got nodes, i.e.
in your example both class definitions end up in one node.  I know that
that's not great readability wise, but the code could have been written

useClass = classA if aTest else classB

with classA and classB being defined at the normal level.

The only reason to define non-trivial classes in an if that I can think of
is machine generated code that probably isn't intended for humans to read.

OTOH if the classes are trivial:

if aTest:
  class Server(TCPServer, MyServer):
    pass
else:
  class Server(UDPServer, MyServer):
    pass

I wouldn't want them split in separate nodes.

Splitting an if across nodes is maybe not the most expected thing.

Are there any in hand examples where the class definitions or defs or
whatever are long enough to make putting them in the same node a major
problem?

Cheers -Terry

**Comment by edreamleo** [comment]
​​

On Wed, Dec 28, 2016 at 10:25 AM, Terry Brown <notifications@github.com>
wrote:

> I wouldn't think badly of Leo if only top level definitions got nodes


​That's the way the old importers work.
​


> the code could have been written
>
> useClass = classA if aTest else classB
>

​That doesn't work if the `if` statement is a guard.  For example, one
class may be a subclass of defined in another file which is imported only
if the guard is True.
​


> Splitting an if across nodes is maybe not the most expected thing.
>

​Alas, it is the *only* thing that could possibly work if long classes are
to be split as expected.​


Are there any in hand examples where the class definitions or defs or
> whatever are long enough to make putting them in the same node a major
> problem?
>

​The original example is from leoColorizer.py.  The problem class is
LeoQtSyntaxHighlighter.  It has only two methods, but imo it should be
represented in a tree of its own.  In Leo's actual source code this is
possible because we use `@file leoColorizer.py`.

A trick will facilitate experimentation.  We can split the problem into two
parts: recognition and code generation.  A new unit test should be the only
test that "triggers" the special case under discussion.  Only after
recognition happens reliably will we need to consider how to generate the
desired nodes.

Furthermore, it may be desirable/necessary to scan the entire problem class
before doing code generation.  This is straightforward with the skip_count
pattern.

In short, I have some hopes for generating the best possible code with a
reasonable amount of effort.

​Edward

**Comment by edreamleo** [comment]
I am a bit daunted by this issue.  Usually when I awake, the solution to the problem I am working on appears.  This morning, however, I awoke with thoughts that no easy solution is possible. That leaves the ugly solutions ;-)

The post pass can't handle this issue because the post pass has no way of knowing which side of the `@others` directive various nodes should be.  Therefore, `python_i.gen_lines` and helpers must properly allocate nodes.  It's likely that `python_i.ends_block` will require substantial revision.  That's too bad: this method is already too ugly.

As stated previously, most importers do not allow section references, and we aren't going to change this.  Even with section references, neither the recognition problem nor the code generation problem would be easy.

The actual generated code is likely to be ugly.  I won't be happy with duplicating what the old python importer does.  The new importer will probably have to handle "if", "else" and "elif" statements as special cases, maybe everywhere, maybe only at the top level. If these statements happen to contain classes or defs, then the importer will actually have to create separate nodes for them.  It's very ugly. So ugly that it may require new invention to be manageable.

**Aside**: Leo can work as an external diff because importers break files into nodes in a consistent way.  If Leo is ever going to replace diff, we can't have python importers ever fail.

**Comment by tbnorth** [comment]
On Thu, Dec 29, 2016 at 4:39 AM, Edward K. Ream <notifications@github.com>
wrote:

> I am a bit daunted by this issue. Usually when I awake, the solution to
> the problem I am working on appears. This morning, however, I awoke with
> thoughts that no easy solution is possible. That leaves the ugly solutions
> ;-)
>
I think that's because the situation is inherently "ugly" :)

I know an editor should support any valid code, but large definitions in
if/then are odd - except for machine generated code, why would they exist?
Are then any examples?  Seems the examples you've given could be addressed
by simply putting everything in one node.  Too much splitting is just as
intrusive as too little splitting.  I often find the splitting of
docstrings / imports from the first node inconvenient, and splitting
__init__ from the class def.

Cheers -Terry

**Comment by edreamleo** [comment]
On Thu, Dec 29, 2016 at 4:11 PM, Terry Brown <notifications@github.com>
wrote:

>
> I know an editor should support any valid code, but large definitions in
> if/then are odd - except for machine generated code, why would they exist?
>

As I have already explained, perhaps not too clearly, the following pattern
is used in Leo's core and in several plugins:

```python
try:
    import aClass
except ImportError:
    aClass = None
...
if aClass:

    class myClass(aClass):
        ...
```
It's no good excusing the importers by saying this pattern is unreasonable.

The old python importer did the bare minimum.  I'd like the new python
importer to do better.

When I awoke this morning I saw that the heart of the problem is an
overindented class definition, that is, a class definition whose
indentation is greater than the cumulative indentations of @others in
generated ancestor nodes.  This is easily calculated, because the top-level
@others has no indentation, while all other @others have one tab-width's
worth of indentation.

Overindented def's are rare, and can be handled more easily.

Too much splitting is just as
> ​ ​
> intrusive as too little splitting. I often find the splitting of
> ​ ​
> docstrings / imports from the first node inconvenient, and splitting
> ​ ​
> __init__ from the class def.
>

​True, but this is a separate, minor, issue. It could be handled in the
post pass.

EKR

**Comment by edreamleo** [comment]
**Status Report**

Experiments yesterday got close, but not close enough.  Just recognizing overindented classes and defs is tricky in the present line-by-line scheme.  I'll give it at least one more shot.  If that doesn't work, the code will have to fully rescan classes and defs.

No matter how classes and defs are parsed, code generation is a problem.  Yesterday I added code to recognize if/else/elif statements.  It's reasonable to create nodes for such when they appear at the top level, but maybe only if they actually contain classes or defs.

Another complication:  Most importers rightly do no not generate section refs, because valid source code could look like a section ref.  But for use by a diff program, we might want to allow section references.  That's for later, of course.  Remember, code generation would be complicated even if section refs were allowed.

All this goes to show how good the old importers (based on basescanner.py) were.  Leo will never again use any of this old code, but they serve as a reference for further work.

**Comment by edreamleo** [comment]
Closed, as of rev 14424de.  Here is the checkin log:

A major milestone.  The python importers correctly import all files in leo/core.
    
The solution was straightforward: starts_block returns False for a few special cases. In those cases, the python importer does *not* split the class into methods. Perhaps it could be done, by creating artificial nodes, but the result would not be Leonine.
